### PR TITLE
Clone opts

### PR DESCRIPTION
**Pull request by bitflower**
Fixes https://github.com/jb55/s3-blob-store/issues/19

@claustres is this a good fix?

### CODE REVIEW

**Comment by jb55** [comment]
looks reasonable to me!

**Comment by bitflower** [comment]
Hey, cool. Looking at it again: doesn‘t this forget to check if opts has a value as in the original solution?

@claustres opinion?

**Comment by claustres** [comment]
`Object.assign()` will not be sufficient because it will only copy top-level references and there are sub-objects like `options.params` that are modified as `options.params.Bucket = xxx`. You need to perform a deep copy to avoid any update-in-place (manually or using eg https://lodash.com/docs/4.17.15#cloneDeep)

**Comment by bitflower** [comment]
Got it, will adapt. Thanks for the input.

**Comment by bitflower** [comment]
Hey guys, I've updated the PR to include @claustres's fix. Tests have passed on my end.

I've looked away from integrating `lodash`'s cloneDeep method and used one from a gist source. If you have better ways let me know!

File: utils/object.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+// Source: https://gist.github.com/Ely-S/4191458
+module.exports = function(obj) {
+  var clone = {};
+  for (var i in obj)
```
Comment (jb55) [inline utils/object.js]: nit: braces around this for block

File: utils/object.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+    clone[i] =
+      obj[i] && typeof obj[i] == 'object'
+        ? x(obj[i].constructor(), obj[i])
```
Comment (jb55) [inline utils/object.js]: what is this x function?

**Comment by jb55** [comment]
fwiw I'm fine with lodash since it's pretty much always pulled into `node_modules` anyways from other dependencies

**Comment by bitflower** [comment]
Right, moved to `lodash.clonedeep` package.

File: utils/object.js (hunk 1 above)
Comment (bitflower) [reply utils/object.js]: obsolete now with `lodash` being used...

File: utils/object.js (hunk 2 above)
Comment (bitflower) [reply utils/object.js]: obsolete now with `lodash` being used...

**Comment by bitflower** [comment]
Not sure how this change would affect travis?

**Comment by jb55** [comment]
I'm going to fix the tests before I merge

**Comment by bitflower** [comment]
hey hey, any chance we can have this soon? Would be awesome!

**Comment by jb55** [comment]
I couldn't fix the tests but this seems fine so I'll merge it

**Comment by jb55** [comment]
merged in 164149dd3d8562c434d022d56551d8115b85c19d, release on 4.1.0

**Pull request closed without merging**